[Lens] Improve transform validation checking

nickofthyme opened the pull request:
## Summary

This PR is meant to improve the Lens transform validation framework.

Closes #242663
Closes #267582

Currently there are a few main problems:
- It's verbose when using over and over for the same chart
- It's split from the api-to-state validator
- It does not ever compare with the initial state when checking state-to-api.

The first 2 are fixed but cleaning up the methods calls.

```ts
// The old way ❌ 
validateConverter(simpleMetricAttributes, metricStateSchema);
validateAPIConverter(simpleMetricApi, metricStateSchema);

// The new way ✅ 
validator.metric.fromState(simpleMetricAttributes);
validator.metric.fromApi(simpleMetricApi);
```

The 3rd issue, the biggest problem is that original configurations can be completely wiped away and the test with still pass. The solution before was to just test the transforms starting with the api then going to state and back with `validateAPIConverter`. However, many of the examples we have, like those from integrations, start from the SO.

The idea is to have a `normalizers` for each chart type that [canonicalizes](https://grokipedia.com/page/Canonicalization) or normalizes the original SO into a form that is comparable to the new SO from the config builder output. This means:
- Replacing uuids with their respective human-readable ids
- Stripping or adding defaults
- Cleaning up properties. Mainly omitting `undefined` values instead of explicitly setting the property (i.e. `"filter": undefined`).

Initially, this will be an **opt-in** parameter on the `validator.fromState` method, then enforced always when all charts are stabilized.

```ts
validator.metric.fromState(simpleMetricAttributes, true); // strict checking enabled
```

In a way this is a bit overkill but I think it is worth it to be able to accurately test input to output 1 to 1, 🍎 to 🍏.

This normalization is does per chart with common logic shared between them. The PR starts with the heatmap charts.

The framework allows both mutating the original SO and the transformed SO (original SO -> api -> transformed SO) to make it easier to handle per case changes. For example converting the original random ids to the human readable ids is easier than vice versa. There is also an `ignore` option to completely ignore comparing given fields, this should be the last case as it's better to normalize to establish the baseline to compare, rather than ignore it entirely. But in some cases it's just too complex to do so.

We can compose these normalizers together for a given chart type to include common normalizers but also to group normalizers but issue, type, context, etc.. In the end joining them together with `mergeNormalizers`.

```ts
const alignLegacyTypes: NormalizerConfig<HeatmapAttributes> = {
  original: (attributes) => {
    // mutate original attributes as needed.
    attributes.state.visualization.legend.isVisible = isVisible ?? true;
    return attributes;
  },
  transformed: (attributes) => {
    // mutate transformed attributes as needed.
    return attributes;
  },
  ignore: ['state.datasourceStates.formBased.layers.*.columns.*.params']
}

export const normalizeHeatmap = mergeNormalizers([alignLegacyTypes])
```

> [!IMPORTANT]
> The normalizers can mutate the `attributes` directly to make it easier to tweak and this is just test code so I don't really care to get fancy. Also the `attributes` should generally not be shared between tests so there should be no interference.

These normalizers **only** apply to the `fromState` validator and only apply when `strict` mode is set to `true` when comparing the original and transformed states.

### Examples

The simplest case is we just compare what we put into the CB against the SO that came out of the CB (SO -> API -> SO).

<details><summary>Simple Metric</summary>
<p>

```diff
- Expected  - 18
+ Received  + 14

@@ -1,61 +1,57 @@
  Object {
    "description": "",
    "references": Array [
      Object {
        "id": "90943e30-9a47-11e8-b64d-95841ca0b247",
-       "name": "indexpattern-datasource-layer-2821bd27-b805-4dea-a7d4-123c248e63b1",
+       "name": "indexpattern-datasource-layer-layer_0",
        "type": "index-pattern",
      },
    ],
    "state": Object {
      "adHocDataViews": Object {},
      "datasourceStates": Object {
        "formBased": Object {
          "layers": Object {
-           "2821bd27-b805-4dea-a7d4-123c248e63b1": Object {
+           "layer_0": Object {
              "columnOrder": Array [
-               "812a7944-731e-4967-8b84-1c8bba4ff04b",
+               "metric_formula_accessor_metric",
              ],
              "columns": Object {
-               "812a7944-731e-4967-8b84-1c8bba4ff04b": Object {
+               "metric_formula_accessor_metric": Object {
+                 "customLabel": false,
                  "dataType": "number",
+                 "filter": undefined,
                  "isBucketed": false,
-                 "label": "Count of records",
+                 "label": "",
                  "operationType": "count",
                  "params": Object {
                    "emptyAsNull": true,
                  },
                  "sourceField": "___records___",
                },
              },
-             "incompleteColumns": Object {},
+             "ignoreGlobalFilters": false,
              "sampling": 1,
-           },
-         },
            },
-       "indexpattern": Object {
-         "layers": Object {},
          },
-       "textBased": Object {
-         "layers": Object {},
        },
      },
      "filters": Array [],
      "internalReferences": Array [],
      "query": Object {
        "language": "kuery",
        "query": "",
      },
      "visualization": Object {
-       "layerId": "2821bd27-b805-4dea-a7d4-123c248e63b1",
+       "collapseFn": undefined,
+       "layerId": "layer_0",
        "layerType": "data",
-       "metricAccessor": "812a7944-731e-4967-8b84-1c8bba4ff04b",
-       "secondaryLabelPosition": "before",
-       "secondaryTrend": Object {
-         "type": "none",
-       },
+       "metricAccessor": "metric_formula_accessor_metric",
+       "showBar": false,
+       "subtitle": "",
+       "valueFontMode": "default",
      },
    },
    "title": "Lens Metric - By Ref",
    "version": 1,
    "visualizationType": "lnsMetric",
```

> Notice many things align but apart from the ids, only a few things are mismatched. 

</p>
</details> 

If we replace all the uuids with their new respective human-readable ids, the true issues become more apparent...

<details><summary>Simple Metric with ids replaced</summary>
<p>

```diff
- Expected  - 12
+ Received  +  8

@@ -16,46 +16,42 @@
              "columnOrder": Array [
                "metric_formula_accessor_metric",
              ],
              "columns": Object {
                "metric_formula_accessor_metric": Object {
+                 "customLabel": false,
                  "dataType": "number",
+                 "filter": undefined,
                  "isBucketed": false,
-                 "label": "Count of records",
+                 "label": "",
                  "operationType": "count",
                  "params": Object {
                    "emptyAsNull": true,
                  },
                  "sourceField": "___records___",
                },
              },
-             "incompleteColumns": Object {},
+             "ignoreGlobalFilters": false,
              "sampling": 1,
            },
-         },
-       },
-       "indexpattern": Object {
-         "layers": Object {},
          },
-       "textBased": Object {
-         "layers": Object {},
        },
      },
      "filters": Array [],
      "internalReferences": Array [],
      "query": Object {
        "language": "kuery",
        "query": "",
      },
      "visualization": Object {
+       "collapseFn": undefined,
        "layerId": "layer_0",
        "layerType": "data",
        "metricAccessor": "metric_formula_accessor_metric",
-       "secondaryLabelPosition": "before",
-       "secondaryTrend": Object {
-         "type": "none",
-       },
+       "showBar": false,
+       "subtitle": "",
+       "valueFontMode": "default",
      },
    },
    "title": "Lens Metric - By Ref",
    "version": 1,
    "visualizationType": "lnsMetric",
```

</p>
</details> 

Going one step further and cleaning up both the common state and the chart-specific visualization state, we see true issues, in this case that the label is cleared and some default options are being applied. We can ignore default options on a per-chart basis while still getting most of the benefits of the comparison. 

<details><summary>Simple Metric with ids replaced</summary>
<p>

```diff
- Expected  - 12
+ Received  +  8

@@ -16,46 +16,42 @@
              "columnOrder": Array [
                "metric_formula_accessor_metric",
              ],
              "columns": Object {
                "metric_formula_accessor_metric": Object {
+                 "customLabel": false,
                  "dataType": "number",
                  "isBucketed": false,
-                 "label": "Count of records",
+                 "label": "",
                  "operationType": "count",
                  "params": Object {
                    "emptyAsNull": true,
                  },
                  "sourceField": "___records___",
                },
              },
               "ignoreGlobalFilters": false,
              "sampling": 1,
            },
        },
      },
      "filters": Array [],
      "internalReferences": Array [],
      "query": Object {
        "language": "kuery",
        "query": "",
      },
      "visualization": Object {
        "layerId": "layer_0",
        "layerType": "data",
        "metricAccessor": "metric_formula_accessor_metric",
-       "secondaryLabelPosition": "before",
-       "secondaryTrend": Object {
-         "type": "none",
-       },
+       "showBar": false,
+       "subtitle": "",
+       "valueFontMode": "default",
      },
    },
    "title": "Lens Metric - By Ref",
    "version": 1,
    "visualizationType": "lnsMetric",
```

</p>
</details> 

### Checklist

Check the PR satisfies following conditions. 

Reviewers should verify this PR satisfies this list as well.

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios
- [x] The PR  description includes the appropriate Release Notes section, and the correct `release_note:*` label is applied per the [guidelines](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)
- [x] Review the [backport guidelines](https://docs.google.com/document/d/1VyN5k91e5OVumlc0Gb9RPa3h1ewuPE705nRtioPiTvY/edit?usp=sharing) and apply applicable `backport:*` labels.

<!--ONMERGE {"backportTargets":["9.4"]} ONMERGE-->

nickofthyme committed:
test: improve transform validation checking

nickofthyme committed:
test: update tests to use new methods

nickofthyme committed:
chore: simplify internal validator logic to implicitly find values per chart type

nickofthyme committed:
chore: add option to ignore fields

nickofthyme committed:
chore: cleanup metric properties

nickofthyme committed:
fix: lens cleared label issue

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
WIP

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
test and type cleanup

nickofthyme committed:
refactor validation logic to improve stability and usability

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
cleanup test logic

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
cleanup types and tests

nickofthyme committed:
minify lens panels json

nickofthyme commented:
/ci

kibanamachine committed:
Changes from node scripts/regenerate_moon_projects.js --update

nickofthyme committed:
fix failing normalizers

nickofthyme committed:
add `position` back to heatmap legend

nickofthyme committed:
remove skipped panels

nickofthyme committed:
make position optional

Co-authored-by: macroscopeapp[bot] <170038800+macroscopeapp[bot]@users.noreply.github.com>

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme commented:
/ci

nickofthyme committed:
add back `excludedFields` on partition tests

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
remove strict validation until metric is normalized

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

kibanamachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/436013)
* Commit: 9651685feb4a5c6cb5ce49c143562f482e1da60f

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/17b7da549d8c568f736b9660ee29a838a1b52b1c) | [after](https://github.com/elastic/kibana/commit/9651685feb4a5c6cb5ce49c143562f482e1da60f) | diff |
| --- | --- | --- | --- |
| `agentBuilder` | 1489 | 1492 | +3 |
| `cases` | 1973 | 1976 | +3 |
| `dashboard` | 1080 | 1083 | +3 |
| `dashboardAgent` | 579 | 582 | +3 |
| `discover` | 2021 | 2024 | +3 |
| `infra` | 1869 | 1872 | +3 |
| `observability` | 1749 | 1752 | +3 |
| `observabilityAIAssistantApp` | 839 | 842 | +3 |
| `unifiedDocViewer` | 914 | 917 | +3 |
| total |  |  | +27 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/17b7da549d8c568f736b9660ee29a838a1b52b1c) | [after](https://github.com/elastic/kibana/commit/9651685feb4a5c6cb5ce49c143562f482e1da60f) | diff |
| --- | --- | --- | --- |
| `@kbn/lens-embeddable-utils` | 158 | 160 | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/17b7da549d8c568f736b9660ee29a838a1b52b1c) | [after](https://github.com/elastic/kibana/commit/9651685feb4a5c6cb5ce49c143562f482e1da60f) | diff |
| --- | --- | --- | --- |
| `agentBuilder` | 1.1MB | 1.1MB | +258.0B |
| `cases` | 2.3MB | 2.3MB | +258.0B |
| `dashboard` | 1.1MB | 1.1MB | +258.0B |
| `dashboardAgent` | 388.3KB | 388.5KB | +258.0B |
| `discover` | 1.9MB | 1.9MB | +258.0B |
| `infra` | 1.5MB | 1.5MB | +258.0B |
| `lens` | 2.0MB | 2.0MB | +401.0B |
| `observability` | 2.0MB | 2.0MB | +255.0B |
| `observabilityAIAssistantApp` | 658.1KB | 658.4KB | +259.0B |
| `unifiedDocViewer` | 672.7KB | 673.0KB | +258.0B |
| total |  |  | +2.7KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/17b7da549d8c568f736b9660ee29a838a1b52b1c) | [after](https://github.com/elastic/kibana/commit/9651685feb4a5c6cb5ce49c143562f482e1da60f) | diff |
| --- | --- | --- | --- |
| `observability` | 97.8KB | 97.8KB | -1.0B |
| `observabilityAIAssistantApp` | 15.7KB | 15.7KB | +1.0B |
| total |  |  | -0.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/17b7da549d8c568f736b9660ee29a838a1b52b1c) | [after](https://github.com/elastic/kibana/commit/9651685feb4a5c6cb5ce49c143562f482e1da60f) | diff |
| --- | --- | --- | --- |
| `@kbn/lens-embeddable-utils` | 183 | 187 | +4 |


</details>


### History
* :broken_heart: [Build #435295](https://buildkite.com/elastic/kibana-pull-request/builds/435295) failed ef99675bab77bdeace3b136c7611eccc9d55e59c
* :broken_heart: [Build #435278](https://buildkite.com/elastic/kibana-pull-request/builds/435278) failed e9f5a0c4acaa929941b978c448ec72b47ee011b4
* :broken_heart: [Build #434514](https://buildkite.com/elastic/kibana-pull-request/builds/434514) failed 5fe98d31025293d83b4d3cf09be8da845fb54a0f

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/436013","number":436013,"commit":"9651685feb4a5c6cb5ce49c143562f482e1da60f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/435295","number":435295,"commit":"ef99675bab77bdeace3b136c7611eccc9d55e59c"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/435278","number":435278,"commit":"e9f5a0c4acaa929941b978c448ec72b47ee011b4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/434514","number":434514,"commit":"5fe98d31025293d83b4d3cf09be8da845fb54a0f"}],"number":436013}
buildkite-pr-comment-->

nickofthyme committed:
compress lens panel json

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
cleanup and move shared code to common util

kibanamachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/436769)
* Commit: b6bfa4abbc8a742170a37e9a6b503164bc485cae

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
#### Module Count
> Fewer modules leads to a faster build time


| id | [before](https://github.com/elastic/kibana/commit/6f32884680529490b96e5c15226ac2f98a398014) | [after](https://github.com/elastic/kibana/commit/b6bfa4abbc8a742170a37e9a6b503164bc485cae) | diff |
| --- | --- | --- | --- |
| `agentBuilder` | 1494 | 1497 | +3 |
| `cases` | 1978 | 1981 | +3 |
| `dashboard` | 1082 | 1085 | +3 |
| `dashboardAgent` | 584 | 587 | +3 |
| `discover` | 2026 | 2029 | +3 |
| `infra` | 1874 | 1877 | +3 |
| `observability` | 1754 | 1757 | +3 |
| `observabilityAIAssistantApp` | 844 | 847 | +3 |
| `unifiedDocViewer` | 919 | 922 | +3 |
| total |  |  | +27 |

#### Public APIs missing comments
> Total count of every public API that lacks a comment. Target amount is 0. Run `node scripts/build_api_docs --plugin [yourplugin] --stats comments` for more detailed information.


| id | [before](https://github.com/elastic/kibana/commit/6f32884680529490b96e5c15226ac2f98a398014) | [after](https://github.com/elastic/kibana/commit/b6bfa4abbc8a742170a37e9a6b503164bc485cae) | diff |
| --- | --- | --- | --- |
| `@kbn/lens-embeddable-utils` | 158 | 160 | +2 |

#### Async chunks
> Total size of all lazy-loaded chunks that will be downloaded as the user navigates the app


| id | [before](https://github.com/elastic/kibana/commit/6f32884680529490b96e5c15226ac2f98a398014) | [after](https://github.com/elastic/kibana/commit/b6bfa4abbc8a742170a37e9a6b503164bc485cae) | diff |
| --- | --- | --- | --- |
| `agentBuilder` | 1.1MB | 1.1MB | +258.0B |
| `cases` | 2.3MB | 2.3MB | +258.0B |
| `dashboard` | 1.1MB | 1.1MB | +258.0B |
| `dashboardAgent` | 388.4KB | 388.6KB | +258.0B |
| `discover` | 1.9MB | 1.9MB | +258.0B |
| `infra` | 1.5MB | 1.5MB | +258.0B |
| `lens` | 2.0MB | 2.0MB | +401.0B |
| `observability` | 2.0MB | 2.0MB | +255.0B |
| `observabilityAIAssistantApp` | 658.7KB | 658.9KB | +258.0B |
| `unifiedDocViewer` | 672.8KB | 673.1KB | +258.0B |
| total |  |  | +2.7KB |

#### Page load bundle
> Size of the bundles that are downloaded on every page load. Target size is below 100kb


| id | [before](https://github.com/elastic/kibana/commit/6f32884680529490b96e5c15226ac2f98a398014) | [after](https://github.com/elastic/kibana/commit/b6bfa4abbc8a742170a37e9a6b503164bc485cae) | diff |
| --- | --- | --- | --- |
| `observability` | 97.8KB | 97.8KB | -1.0B |

<details><summary>Unknown metric groups</summary>

#### API count
| id | [before](https://github.com/elastic/kibana/commit/6f32884680529490b96e5c15226ac2f98a398014) | [after](https://github.com/elastic/kibana/commit/b6bfa4abbc8a742170a37e9a6b503164bc485cae) | diff |
| --- | --- | --- | --- |
| `@kbn/lens-embeddable-utils` | 183 | 187 | +4 |


</details>


### History
* :yellow_heart: [Build #436744](https://buildkite.com/elastic/kibana-pull-request/builds/436744) was flaky 3fd924016aad8a3dc30d685e9bd98fa27e3efc78
* :green_heart: [Build #436013](https://buildkite.com/elastic/kibana-pull-request/builds/436013) succeeded 9651685feb4a5c6cb5ce49c143562f482e1da60f
* :broken_heart: [Build #435295](https://buildkite.com/elastic/kibana-pull-request/builds/435295) failed ef99675bab77bdeace3b136c7611eccc9d55e59c
* :broken_heart: [Build #435278](https://buildkite.com/elastic/kibana-pull-request/builds/435278) failed e9f5a0c4acaa929941b978c448ec72b47ee011b4
* :broken_heart: [Build #434514](https://buildkite.com/elastic/kibana-pull-request/builds/434514) failed 5fe98d31025293d83b4d3cf09be8da845fb54a0f

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/436769","number":436769,"commit":"b6bfa4abbc8a742170a37e9a6b503164bc485cae"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/436744","number":436744,"commit":"3fd924016aad8a3dc30d685e9bd98fa27e3efc78"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/436013","number":436013,"commit":"9651685feb4a5c6cb5ce49c143562f482e1da60f"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/435295","number":435295,"commit":"ef99675bab77bdeace3b136c7611eccc9d55e59c"},{"buildStatus":{"state":"failing","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/435278","number":435278,"commit":"e9f5a0c4acaa929941b978c448ec72b47ee011b4"},{"buildStatus":{"state":"failed","success":false,"hasRetries":true,"hasNonPreemptionRetries":true},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/434514","number":434514,"commit":"5fe98d31025293d83b4d3cf09be8da845fb54a0f"}],"number":436769}
buildkite-pr-comment-->

nickofthyme committed:
generate panel title when empty

andrimal commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/tests/integrations/so_comparison.ts:
```diff
@@ -0,0 +1,269 @@
+ * - Rewriting layer and column IDs to match generated convention
+ */
+export function reshapeAttributes(attributes: LensAttributes): LensAttributes {
+  const ACCESSOR_TYPE_MAP: AccessorMapping = {
```
I believe this `reshapeAttributes` function is not currently used anywhere. Should we remove the whole file in favor of the `getCommonNormalizer` now?

nickofthyme replied:
Yeah thanks this was Marco's first pass solution. Removed in f111765.

andrimal commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/tests/utils/unset_many.ts:
```diff
@@ -0,0 +1,18 @@
+  paths.forEach((path) => unset(result, path));
+
+  return result;
+}
```
Same here, I believe this function is not used anymore. :)

nickofthyme replied:
Thanks, removed in f111765

andrimal commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/tests/utils/validator/normalizers/normalize.ts:
```diff
@@ -0,0 +1,125 @@
+  ignore: string[] = []
+): AttributesNormalizer<T> {
+  return function (attributes: NormalizedStates<T>): NormalizedStates<T> {
+    const sortedNormalizers = normalizers.sort((a, b) => (a.order ?? 0) - (b.order ?? 0));
```
I think sort mutates the `normalizers` directly:

```suggestion
    const sortedNormalizers = [...normalizers].sort((a, b) => (a.order ?? 0) - (b.order ?? 0));
```

nickofthyme replied:
Probably best not to mutate, f111765.

andrimal commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/tests/xy/xy.test.ts:
```diff
@@ -8,12 +8,12 @@
 
 import type { XYVisualizationState } from '@kbn/lens-common';
-import { xyConfigSchema } from '../../schema/charts/xy';
+import {} from '../../schema/charts/xy';
```
we could remove this empty `import` :)

nickofthyme replied:
Removed in f111765 😅

andrimal commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/tests/utils/validator/validate_state_transforms.ts:
```diff
@@ -0,0 +1,61 @@
+  const schema = getChartSchema(chartType);
+  const builder = new LensConfigBuilder(undefined, true);
+
+  return function validateStateTransforms(attributes, strict = false, excludedFields = []) {
```
`excludedFields` are not used in the function for now. Maybe we could omit it?

nickofthyme replied:
Removed in 06fd847

andrimal reviewed (commented):
Great work @nickofthyme! ✨ I only left some cleaning-up comments.

nickofthyme commented on src/platform/packages/shared/kbn-lens-embeddable-utils/config_builder/tests/integrations/lens_panels.json.gz:
@markov00 good idea on the compressing, gz is much nicer! 200x smaller with no impact to test run.

| | json | gz |
|-|-|-|
| File size | 41.2 MB (~21 MB compact) | 1.99 MB |
| Test run (`--no-cache`) | 67.9 s | 66.2 s |

nickofthyme committed:
Fix common tranform issues
- Normalize `textBased` data sources
- Normalize formula based columns, add example
- Set default panel title
- Enabled strict check for all heatmaps
- add heatmap SO version

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
code cleanup from review comments

nickofthyme committed:
normalize to latest version

nickofthyme committed:
add notes on conflicting props

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
Merge branch 'main' into lens-transform-validation

nickofthyme committed:
Merge branch 'main' into lens-transform-validation
